core/internal/runfiles/upload_batcher.go

```diff
@@ -0,0 +1,83 @@
+package runfiles
+
+import (
+	"slices"
+	"sync"
+)
+
+// uploadBatcher helps batch many simultaneous upload operations.
+type uploadBatcher struct {
+	sync.Mutex
+
+	// Wait group for Add-ed files to get uploaded.
+	addWG *sync.WaitGroup
+
+	// Files collected so far.
+	files []string
+
+	// Whether an upload is queued to happen soon.
+	isQueued bool
+
+	// A function that returns a stream that is signalled when the next
+	// batch upload should happen.
+	delayFunc func() <-chan struct{}
+
+	// Callback to upload a list of files.
+	upload func([]string)
+}
+
+func newUploadBatcher(
+	delayFunc func() <-chan struct{},
+	upload func([]string),
+) *uploadBatcher {
+	return &uploadBatcher{
+		addWG: &sync.WaitGroup{},
+		files: make([]string, 0),
+
+		delayFunc: delayFunc,
+		upload:    upload,
+	}
+}
+
+// Add adds files to the next upload batch, scheduling one if necessary.
+func (b *uploadBatcher) Add(files []string) {
+	if b.delayFunc == nil {
+		b.upload(files)
+		return
+	}
+
+	b.Lock()
+	defer b.Unlock()
+
+	b.files = append(b.files, files...)
+
+	if !b.isQueued {
+		b.addWG.Add(1)
+		b.isQueued = true
+
+		go b.uploadAfterDelay()
+	}
+}
+
+// Finish blocks until all files from previous Add calls are uploaded.
+func (b *uploadBatcher) Finish() {
+	b.addWG.Wait()
+}
+
+func (b *uploadBatcher) uploadAfterDelay() {
+	<-b.delayFunc()
+
+	b.Lock()
+	b.isQueued = false
+	files := b.files
+	b.files = make([]string, 0)
+	b.Unlock()
+
+	// Deduplicate.
+	slices.Sort(files)
+	files = slices.Compact(files)
```

Comment:
if you use a "set" i.e. a map with a struct you don't need to de-dup, it will already be de-dupped, but you will have to pay the conversion to a list i guess.

Reply:
great point, done!